crates/xline-client/src/clients/lock.rs

```diff
@@ -1,78 +1,242 @@
+    RangeResponse, Request, RequestOp, RequestWrapper, Response, ResponseHeader, SortOrder,
+    SortTarget, TargetUnion, TxnRequest, TxnResponse,
 };
 
 use crate::{
-    clients::{lease::LeaseClient, watch::WatchClient},
+    clients::{lease::LeaseClient, watch::WatchClient, DEFAULT_SESSION_TTL},
     error::{Result, XlineClientError},
     lease_gen::LeaseIdGenerator,
     types::{
-        lease::LeaseGrantRequest,
-        lock::{LockRequest, UnlockRequest},
+        lease::{LeaseGrantRequest, LeaseKeepAliveRequest},
         watch::WatchRequest,
     },
     CurpClient,
 };
 
-/// Client for Lock operations.
-#[derive(Clone)]
-pub struct LockClient {
-    /// The client running the CURP protocol, communicate with all servers.
-    curp_client: Arc<CurpClient>,
-    /// The lease client
-    lease_client: LeaseClient,
-    /// The watch client
-    watch_client: WatchClient,
-    /// Auth token
-    token: Option<String>,
+/// Session represents a lease kept alive for the lifetime of a client.
+#[derive(Debug)]
+pub struct Session {
+    /// The lock client that used to create the session
+    client: LockClient,
+    /// lease id
+    lease_id: i64,
+    /// lease lifetime
+    ttl: i64,
+    /// `keep_alive` task will auto-renew the lease
+    keep_alive: Option<JoinHandle<Result<()>>>,
 }
 
-impl Debug for LockClient {
+impl Session {
+    /// Creates a new session.
     #[inline]
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("LockClient")
-            .field("lease_client", &self.lease_client)
-            .field("watch_client", &self.watch_client)
-            .field("token", &self.token)
-            .finish()
+    #[must_use]
+    pub fn new(client: LockClient) -> Self {
+        Self {
+            client,
+            lease_id: 0,
+            ttl: DEFAULT_SESSION_TTL,
+            keep_alive: None,
+        }
+    }
+
+    /// Set session TTL.
+    #[inline]
+    #[must_use]
+    pub const fn with_ttl(mut self, ttl: i64) -> Self {
+        self.ttl = ttl;
+        self
+    }
+
+    /// Set session lease.
+    #[inline]
+    #[must_use]
+    pub const fn with_lease(mut self, lease_id: i64) -> Self {
+        self.lease_id = lease_id;
+        self
+    }
+
+    /// Crate an auto-renew session.
+    ///
+    /// # Errors
+    /// return errors when lease client grant failed
+    #[inline]
+    pub async fn build(mut self) -> Result<Self> {
+        let ttl = self.ttl;
+        self.lease_id = if self.lease_id == 0 {
+            let lease_response = self
+                .client
+                .lease_client
+                .grant(LeaseGrantRequest::new(ttl))
+                .await?;
+            lease_response.id
+        } else {
+            self.lease_id
+        };
+        let mut lease_client = self.client.lease_client.clone();
+        let lease_id = self.lease_id;
+        self.keep_alive = Some(tokio::spawn(async move {
```

Comment:
Done

Reply:
It seems we can do nothing but panic. Do you have any better solutions?

Reply:
On second thought, I think it's impossible to make sure that the lock is always held during crital section, the user must assume the lock is held.

Reply:
I think we could simply print an error message when the task exits.

Reply:
We may consider implementing this https://github.com/xline-kv/Xline/pull/820#issuecomment-2174925161

Reply:
Currently, a user will get an error when he invokes the `lock` method on a `Xutex` of which keep-alive task exists.

Reply:
I suggest to use the following API to avoid confusion about using the lock:
```rust
    let xutex = Xutex::new(..).await?;
    xutex.put(..).await?;
    xutex.range(..).await?;
    xutex.delete_range(..).await?;
    xutex.txn(..).await?;

    let guard = xutex.lock_unsafe().await?;
    // perform some unsafe operation that may break consistency
    // ...
    drop(guard);

```
We could directly implement safe operations `put/range/delete_range/txn` for the type `Xutex`, and also implement a `lock_unsafe` method to make it compatible with the behavior of the etcd client.

Reply:
What does ‘compatible with the behavior of the etcd client’  mean? Just renaming `lock` as `lock_unsafe`?  In addition, I don't get the point why implementing `put/range/delete_range` is less confusing than not implementing them? The `Xutex` is a lock wrapper, not a `KvClient` wrapper. Therefore, if a user needs to do something while holding a lock, he should use a `KvClient` to initiate a request instead of using `Xutex`.  Considering that, I think providing `txn_with_locked_key` is enough. What it does is only to generate a txn request which compares the locked key exist or not.

Reply:
The user should be provided with safe API by default, which are `put/range/delete_range` coupled with the lock key. If the user want the same behavior as etcd, they could use `lock_unsafe`.

Reply:
I suggest not using the term `lock`

Reply:
If the `put/range/delete_range` should coupled with the locked key, that means all they need is a txn. Therefore, using `txn_with_locked_key` is enough.

Reply:
Renamed